spring-security/src/test/java/com/sap/cloud/security/spring/config/IdentityServicesPropertySourceFactoryBrokerNoHoleTest.java

```diff
@@ -5,38 +5,68 @@
+import com.sap.cloud.security.config.OAuth2ServiceConfiguration;
+import com.sap.cloud.security.config.Service;
+import com.sap.cloud.security.config.ServiceBindingEnvironment;
+import com.sap.cloud.security.config.ServiceConstants;
+import org.apache.commons.io.IOUtils;
+import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.Test;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.beans.factory.annotation.Value;
 import org.springframework.boot.test.context.SpringBootTest;
 import org.springframework.context.annotation.Configuration;
 import org.springframework.context.annotation.PropertySource;
 
+/**
+ * Tests that the {@link IdentityServicesPropertySourceFactory} puts 2 XSUAA service instances with plan 'application' into the Spring properties without creating a hole at index 1.
+ * For backward-compatibility, the order of the service instance must be as follows:
+ * Index 0: Configuration accessible via Environment#getXsuaaConfiguration (Application)
+ * Index 1: Configuration accessible via Environment#getXsuaaConfigurationForTokenExchange (Broker) if exists, otherwise next XSUAA configuration
+ * Index 2+: Remaining XSUAA configurations
+ * In addition, tests that the IAS service instance from the environment is correctly added as well.
+ */
 @SpringBootTest(classes = { BrokerHoleTestConfigurationFromFile.class })
 class IdentityServicesPropertySourceFactoryBrokerNoHoleTest {
 
 	@Autowired
 	BrokerHoleTestConfigurationFromFile configuration;
 
+	static String serviceBindingJson;
+
+	@BeforeAll
+	static void setup() throws IOException {
+		serviceBindingJson = IOUtils.resourceToString("/xsuaaBindingsTwoApplicationsNoBroker.json", UTF_8);
+	}
+
 	@Test
-	void testInjectedPropertyValues_fourXsuaaBindings() {
+	void testInjectedPropertyValues() {
+		ServiceBindingEnvironment env = new ServiceBindingEnvironment(new SapVcapServicesServiceBindingAccessor(any -> serviceBindingJson));
+
 		/* Index 0 */
-		assertEquals("client-id2", configuration.xsuaaClientId0);
-		assertEquals("client-secret2", configuration.xsuaaClientSecret0);
-		assertEquals("http://domain.xsuaadomain", configuration.xsuaaUrl0);
-		assertEquals("xsuaadomain", configuration.xsuaaDomain0);
-		assertEquals("xsappname2", configuration.xsuaaAppName0);
+		OAuth2ServiceConfiguration xsuaaConfig = env.getXsuaaConfiguration();
+		assertEquals(xsuaaConfig.getClientId(), configuration.xsuaaClientId0);
+		assertEquals(xsuaaConfig.getClientSecret(), configuration.xsuaaClientSecret0);
+		assertEquals(xsuaaConfig.getProperty(ServiceConstants.URL), configuration.xsuaaUrl0);
+		assertEquals(xsuaaConfig.getProperty(ServiceConstants.XSUAA.UAA_DOMAIN), configuration.xsuaaDomain0);
+		assertEquals(xsuaaConfig.getProperty(ServiceConstants.XSUAA.APP_ID), configuration.xsuaaAppName0);
+		assertEquals(ServiceConstants.Plan.APPLICATION, ServiceConstants.Plan.from(configuration.xsuaaPlan0));
```

Comment:
With checking against `xsuaaConfig` your unit test does not longer test `IdentityServicesPropertySourceFactory` in isolation. Instead, you are also testing `env.getXsuaaConfiguration()` as well. This makes this test already something like a "(module) integration test".
I don't know the detailed guidelines in your project, that is why I am unsure if this is acceptable.

Reply:
Good point. My intention was to use this opportunity to test the backward-compatibility aspect that `IdentityServicesPropertySourceFactory` has to respect since it contains more than two configurations.

I see three options how to resolve it. Which one do you find best?

1. Changing the `IdentityServicesPropertySourceFactory` tests to assert **both** against the string values and against env.getXsuaaConfiguration()?

2. Accept the backward-compatibility aspect as sufficiently tested in `XsuaaServiceConfigurationsTest` where `IdentityServicesPropertySourceFactory` is used to populate the Spring properties. Than again, that test has the same problem you describe.

3. Add a dedicated (module) integration test and revert the `IdentityServicesPropertySourceFactory` tests and `XsuaaServiceConfiguration(s)Test` to test in isolation. Do you have a suggestion for a good name?

Reply:
@liga-oz I have option 3 committed locally in case we want to go with this. It was not that much work since we had the tests in the git history already with/without isolated testing by now. Let me know if I should push when you have time to review.

Reply:
lets go with the option 3 it's my preferred option too and as you already done the work ,makes even more sense to go with it 😄